hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/ContractMintHTSV2SecurityModelSuite.java

```diff
@@ -743,6 +767,181 @@ final HapiSpec V2Security035TokenWithDelegateContractKeyCanNotMintFromDelegateca
+                        emptyChildRecordsCheck(
+                                STATIC_CALL_WHEN_NON_FUNGIBLE_TOKEN_HAS_CONTRACT_ID, CONTRACT_REVERT_EXECUTED));
+    }
+
+    @HapiTest
+    final HapiSpec V2Security040TokenWithDelegateContractKeyCanNotMintFromCallcode() {
+        final AtomicReference<TokenID> fungible = new AtomicReference<>();
+        final AtomicReference<TokenID> nonFungible = new AtomicReference<>();
+
+        return propertyPreservingHapiSpec("V2Security040TokenWithDelegateContractKeyCanNotMintFromCallcode")
+                .preserving(CONTRACTS_MAX_NUM_WITH_HAPI_SIGS_ACCESS)
+                .given(
+                        overriding(CONTRACTS_MAX_NUM_WITH_HAPI_SIGS_ACCESS, CONTRACTS_V2_SECURITY_MODEL_BLOCK_CUTOFF),
+                        cryptoCreate(TOKEN_TREASURY).balance(THOUSAND_HBAR),
+                        cryptoCreate(RECEIVER),
+                        cryptoCreate(SIGNER).balance(THOUSAND_HBAR),
+                        tokenCreate(FUNGIBLE_TOKEN)
+                                .tokenType(TokenType.FUNGIBLE_COMMON)
+                                .initialSupply(0)
+                                .treasury(TOKEN_TREASURY)
+                                .adminKey(TOKEN_TREASURY)
+                                .supplyKey(TOKEN_TREASURY)
+                                .exposingCreatedIdTo(idLit -> fungible.set(asToken(idLit))),
+                        tokenCreate(NON_FUNGIBLE_TOKEN)
+                                .tokenType(TokenType.NON_FUNGIBLE_UNIQUE)
+                                .initialSupply(0)
+                                .treasury(TOKEN_TREASURY)
+                                .adminKey(TOKEN_TREASURY)
+                                .supplyKey(TOKEN_TREASURY)
+                                .exposingCreatedIdTo(idLit -> nonFungible.set(asToken(idLit))),
+                        uploadInitCode(MINT_TOKEN_VIA_CALLCODE),
+                        contractCreate(MINT_TOKEN_VIA_CALLCODE))
+                .when(withOpContext((spec, opLog) -> allRunFor(
+                        spec,
+                        newKeyNamed(CONTRACT_KEY)
+                                .shape(TRESHOLD_KEY_SHAPE.signedWith(sigs(ON, MINT_TOKEN_VIA_CALLCODE))),
+                        tokenUpdate(FUNGIBLE_TOKEN).supplyKey(CONTRACT_KEY),
+                        cryptoUpdate(TOKEN_TREASURY).key(CONTRACT_KEY),
+                        // Test Case 1: Treasury account paying and signing a fungible TOKEN MINT TRANSACTION,
+                        // when the token is expected to be minted in the token treasury account
+                        // SIGNER -> call -> CONTRACT -> callcode -> PRECOMPILE
+                        contractCall(
+                                        MINT_TOKEN_VIA_CALLCODE,
+                                        "callCodeToContractWithoutAmount",
+                                        asHeadlongAddress("0000000000000000000000000000000000000167"),
```

Comment:
Can we extract this address as a constant in this file or in a helper class?

Reply:
nice catch! It is extracted now